internal/server/iigointernal/orchestration.go

```diff
@@ -76,64 +91,82 @@ func RunIIGO(g *gamestate.GameState, clientMap *map[shared.ClientID]baseclient.C
-
 	// 1 Judge actions - inspect history
 	_, _ = judicialBranch.inspectHistory(g.IIGOHistory)
 
 	// 2 President actions
-	resourceReports := map[shared.ClientID]shared.Resources{}
-	var aliveClientIds []shared.ClientID
-	for _, v := range rules.VariableMap[rules.IslandsAlive].Values {
-		aliveClientIds = append(aliveClientIds, shared.ClientID(int(v)))
-		resourceReports[shared.ClientID(int(v))] = iigoClients[shared.ClientID(int(v))].ResourceReport()
+	resourceReports := map[shared.ClientID]shared.ResourcesReport{}
+	aliveClientIds := []shared.ClientID{}
+	for clientID, clientGameState := range g.ClientInfos {
+		if clientGameState.LifeStatus != shared.Dead {
+			aliveClientIds = append(aliveClientIds, clientID)
+			resourceReports[clientID] = iigoClients[clientID].ResourceReport()
+		}
+	}
+
+	// Throw error if any of the actions returns error
+	insufficientBudget := executiveBranch.broadcastTaxation(resourceReports, aliveClientIds)
+	var ruleToVoteReturn shared.PresidentReturnContent
+
+	if insufficientBudget == nil {
+		insufficientBudget = executiveBranch.requestAllocationRequest(aliveClientIds)
+	}
+	if insufficientBudget == nil {
+		insufficientBudget = executiveBranch.replyAllocationRequest(g.CommonPool, aliveClientIds)
+	}
+	if insufficientBudget == nil {
+		insufficientBudget = executiveBranch.requestRuleProposal(aliveClientIds)
+	}
+	if insufficientBudget == nil {
+		ruleToVoteReturn, insufficientBudget = executiveBranch.getRuleForSpeaker()
+	}
+	if insufficientBudget != nil {
+		return false, "Common pool resources insufficient for executiveBranch actions"
 	}
-	executiveBranch.broadcastTaxation(resourceReports)
-	executiveBranch.requestAllocationRequest()
-	executiveBranch.replyAllocationRequest(g.CommonPool)
-	executiveBranch.requestRuleProposal()
-	ruleToVote, ruleSelected := executiveBranch.getRuleForSpeaker()
 
 	// 3 Speaker actions
-	if ruleSelected {
-		legislativeBranch.setRuleToVote(ruleToVote)
-		legislativeBranch.setVotingResult(aliveClientIds)
-		legislativeBranch.announceVotingResult()
+
+	//TODO:- shouldn't updateRules be called here?
+	if ruleToVoteReturn.ActionTaken && ruleToVoteReturn.ContentType == shared.PresidentRuleProposal {
+		insufficientBudget := legislativeBranch.setRuleToVote(ruleToVoteReturn.ProposedRule)
+		if insufficientBudget == nil {
+			insufficientBudget = legislativeBranch.setVotingResult(aliveClientIds)
+		}
+		if insufficientBudget == nil {
+			insufficientBudget = legislativeBranch.announceVotingResult()
+		}
+		if insufficientBudget != nil {
+			return false, "Common pool resources insufficient for legislativeBranch actions"
+		}
 	}
 
+	// TODO:- at the moment, these are action (and cost resources) but should they?
+	var appointJudgeError, appointSpeakerError, appointPresidentError error
 	// Get new Judge ID
-	g.JudgeID = legislativeBranch.appointNextJudge(g.JudgeID, aliveClientIds)
+	g.JudgeID, appointJudgeError = legislativeBranch.appointNextJudge(g.JudgeID, aliveClientIds)
```

Comment:
If the appointNextJudge function fails the current speaker becomes the new judge, is that intended?

Reply:
Yeah.. Originally this wasn't a problem as returning the error meant that non of the roles would change. This has apparently got lost in one of the merges but should be fixed again now.